src/ol/events/eventtarget.js

```diff
@@ -0,0 +1,137 @@
+goog.provide('ol.events.EventTarget');
+
+goog.require('goog.Disposable');
+goog.require('ol.events');
+goog.require('ol.events.Event');
+
+/**
+ * @classdesc
+ * A simplified implementation of the W3C DOM Level 2 EventTarget interface.
+ * @see {@link https://www.w3.org/TR/2000/REC-DOM-Level-2-Events-20001113/events.html#Events-EventTarget}
+ *
+ * There are two important simplifications compared to the specification:
+ *
+ * 1. The handling of `useCapture` in `addEventListener` and
+ *    `removeEventListener`. There is no real capture model. Instead, when
+ *    adding a listener, `useCapture` means that it will be added as first
+ *    listener, causing it to be called before other listeners. When removing a
+ *    listener, the `useCapture` argument will be ignored, and the listener will
+ *    be removed regardless of whether it was added with `useCapture` set to
+ *    true or false.
+ * 2. The handling of `stopPropagation` and `preventDefault` on `dispatchEvent`.
+ *    There is no event target hierarchy. When a listener calls
+ *    `stopPropagation` or `preventDefault` on an event object, it means that no
+ *    more listeners after this one will be called. Same as when the listener
+ *    returns false.
+ *
+ * @constructor
+ * @extends {goog.Disposable}
+ */
+ol.events.EventTarget = function() {
+
+  goog.base(this);
+
+  /**
+   * @private
+   * @type {!Object.<string, Array.<ol.events.ListenerFunctionType>>}
+   */
+  this.listeners_ = {};
+
+};
+goog.inherits(ol.events.EventTarget, goog.Disposable);
+
+
+/**
+ * @param {ol.events.EventType|string} type Type.
+ * @param {ol.events.ListenerFunctionType} listener Listener.
+ * @param {boolean=} opt_capture Call listener before already registered
+ *     listeners. Default is false.
+ */
+ol.events.EventTarget.prototype.addEventListener = function(
+    type, listener, opt_capture) {
+  var listeners = this.listeners_[type];
+  if (!listeners) {
+    listeners = this.listeners_[type] = [];
+  }
+  if (listeners.indexOf(listener) === -1) {
+    if (opt_capture) {
+      listeners.push(listener);
+    } else {
+      listeners.unshift(listener);
+    }
+  }
+};
+
+
+/**
+ * @param {{type: (ol.events.EventType|string),
+ *     target: (EventTarget|ol.events.EventTarget|undefined)}|ol.events.Event|
+ *     ol.events.EventType|string} event Event or event type.
+ * @return {boolean|undefined} `false` if anyone called preventDefault on the
+ *     event object or if any of the listeners returned false.
+ */
+ol.events.EventTarget.prototype.dispatchEvent = function(event) {
+  var evt = goog.isString(event) ? new ol.events.Event(event) : event;
+  var type = evt.type;
+  evt.target = this;
+  var listeners = this.listeners_[type];
+  if (listeners) {
+    for (var i = listeners.length - 1; i >= 0; --i) {
```

Comment:
No, because listeners are added with `unshift()`, not `push()`. Looping in reverse order allows listeners to remove themselves or other listeners without disturbing the loop.